File: threeSum.cpp
//
//  leetcode.com： 15. 3Sum
//

#include <vector>

using namespace std;

class Solution
{
  public:
    vector<vector<int>> threeSum(vector<int> &nums)
    {
        vector<vector<int>> answers;

        if (nums.size() < 3)
            return answers;

        sort(nums.begin(), nums.end());

        int ans = 0;

        for (int i = 0; i < nums.size() - 2; i++)
        {
            if (i > 0 && nums[i] == nums[i - 1])
                continue;

            int j = i + 1;

            int k = nums.size() - 1;

            while (j < k)
            {
                ans = nums[i] + nums[j] + nums[k];

                if (ans == 0)
                {
                    answers.push_back({nums[i], nums[j], nums[k]});
                    ++j;

                    while (j < nums.size() && nums[j] == nums[j - 1])
                        ++j;
                    --k;
                    while (k >= 0 && nums[k] == nums[k + 1])
                        --k;
                }

                else if (ans > 0)
                    --k;

                else
                    ++j;
            }
        }

        return answers;
    }
};

File: leetcode/java/src/main/java/com/vkeonline/leetcode/p000/ThreeSum.java
package com.vkeonline.leetcode.p000;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;


/**
 * Leetcode [M]: 15. 3Sum
 * @author csgear
 */
public class ThreeSum {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> result = new ArrayList<>() ;

        Arrays.sort(nums);

        // nums in ascending order, nums[i] shall be zero or negative as the first in the tuple
        for(int i = 0 ; i < nums.length && nums[i] <=0 ; i++) {
            if(i == 0 || nums[i-1] != nums[i]) {
                twoSum(nums, i, result) ;
            }
        }

        return result ;
    }

    private void twoSum(int[] nums, int current, List<List<Integer>> result) {
        int low = current + 1, high = nums.length - 1 ;
        while(low < high) {
            int sum = nums[current] + nums[low] + nums[high];
            if(sum < 0) {
                low++ ;
            }
            else if(sum > 0) {
                high-- ;
            }
            else {
                result.add(Arrays.asList(nums[current], nums[low], nums[high])) ;
                // used only once to remove duplicated solution
                low++ ;
                high-- ;
                while (low < high && nums[low - 1] == nums[low]) {
                    low++ ;
                }
            }
        }
    }

}
